examples/userapi/mpi.ipynb

```diff
@@ -0,0 +1,1521 @@
+      "outputs": [],
+      "received": "2019-05-10T07:06:36.457184",
+      "started": "2019-05-10T07:06:36.435751",
+      "status": "ok",
+      "stderr": "",
+      "stdout": "",
+      "submitted": "2019-05-10T07:06:36.428798"
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "%%px\n",
+    "u.data_with_halo[:] = 1.\n",
+    "u.data_with_halo"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## MPI and SparseFunction\n",
+    "\n",
+    "A `SparseFunction` represents a sparse set of points which are generically unaligned with the `Grid`. A sparse point could be anywhere within a grid, and is therefore attached some coordinates. Given a sparse point, Devito looks at its coordinates and, based on the domain decomposition, **logically** assigns it to a given MPI process; this is purely logical ownership, as in Python-land, before running an Operator, the sparse point physically lives on the MPI rank which created it. Within `op.apply`, right before jumping to C-land, the sparse points are scattered to their logical owners; upon returning to Python-land, the sparse points are gathered back to their original location.\n",
+    "\n",
+    "In the following example, we attempt injection of four sparse points into the neighboring grid points via linear interpolation."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%px\n",
+    "from devito import Function, SparseFunction\n",
+    "grid = Grid(shape=(4, 4), extent=(3.0, 3.0))\n",
+    "x, y = grid.dimensions\n",
+    "f = Function(name='f', grid=grid)\n",
+    "coords = [(0.5, 0.5), (1.5, 2.5), (1.5, 1.5), (2.5, 1.5)]\n",
+    "sf = SparseFunction(name='sf', grid=grid, npoint=len(coords), coordinates=coords)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let:\n",
+    "* O be a grid point\n",
+    "* x be a halo point\n",
+    "* A, B, C, D be the sparse points\n",
+    "\n",
+    "We show the global view, that is what the user \"sees\".\n",
+    "\n",
+    "```\n",
+    "O --- O --- O --- O\n",
+    "|  A  |     |     |\n",
+    "O --- O --- O --- O\n",
+    "|     |  C  |  B  |\n",
+    "O --- O --- O --- O\n",
+    "|     |  D  |     |\n",
+    "O --- O --- O --- O\n",
+    "```\n",
+    "\n",
+    "And now the local view, that is what the MPI ranks own when jumping to C-land.        \n",
+    "\n",
+    "```\n",
+    "Rank 0          Rank 1\n",
+    "O --- O --- x   x --- O --- O\n",
+    "|  A  |     |   |     |     |\n",
+    "O --- O --- x   x --- O --- O\n",
+    "|     |  C  |   |  C  |  B  |\n",
+    "x --- x --- x   x --- x --- x\n",
+    "\n",
+    "Rank 2           Rank 3\n",
+    "x --- x --- x   x --- x --- x\n",
+    "|     |  C  |   |  C  |  B  |\n",
+    "O --- O --- x   x --- O --- O\n",
+    "|     |  D  |   |  D  |     |\n",
+    "O --- O --- x   x --- O --- O\n",
```

Comment:
a diagram like that could work very well for helping the reader on the beginning of this notebook for a non-sparse case

Reply:
I **think** the ipyparallel output is quite clear...each process has a 2x2 local section of the grid...no?